### PR TITLE
Async kin + Typing

### PR DESCRIPTION
**Pull request by Ronserruya**
Update the kin-sdk to work with asyncio:

Changes:
1. Most of the methods are now async
2. Minimum supported python version changed to 3.6
3. User-agent on http requests changed from "kin-core-python" to "kin-sdk-python"
4. Added type hints to all methods
5. Builder and Horizon modules were removed since we use kin-base instead of inheriting them
6. KinAccount doesn't throw an error when initialized with account that doesn't exist on the blockchain
7. Monitors were changed to async generators and not generators in a different threads
8. New parameter to set timeout between events in payment monitors
9. MultiMonitor was changed to remove the "add/remove_address" methods, the passed set can just be changed at will.
10. Timeout parameter removed from get_channel method, as async queue doesn't provide timeout
11. Added context manager to the KinClient object to close the connection to the blockchain
12. Removed "verify_kin_payment" method
13. Transactions are now retried with the same tx, instead of building a new one each time

### CODE REVIEW

File: images/docker-compose.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ version: "3"
 
 services:
   stellar-core-1:
-    image: kinecosystem/stellar-core:kinecosystem-v2.0.0-stellar-v9.2.0
+    image: kinecosystem/stellar-core:latest
```
Comment (mrsegev) [inline images/docker-compose.yml]: It's better to hardcode the version. This is the only way to be 100% sure the build is stable. Future updates might break the code.

File: images/docker-compose.yml [hunk 2]
```diff
@@ -26,7 +26,7 @@ services:
       POSTGRES_DB: core
 
   stellar-core-2:
-    image: kinecosystem/stellar-core:kinecosystem-v2.0.0-stellar-v9.2.0
+    image: kinecosystem/stellar-core:latest
```
Comment (mrsegev) [inline images/docker-compose.yml]: Same. Let's hardcode the versions in this file.

File: requirements-dev.txt [hunk 3]
```diff
@@ -1,8 +1,5 @@
-attrs==17.4.0
-codecov==2.0.15
-coverage==4.5.1
-funcsigs==1.0.2
-pluggy==0.6.0
-py==1.5.2
-pytest-cov==2.5.1
-pytest==3.4.0
+codecov
```
Comment (mrsegev) [inline requirements-dev.txt]: Let's lock the version here.

File: images/docker-compose.yml (hunk 2 above)
Comment (Ronserruya) [reply images/docker-compose.yml]: blockchain team removes their docker images a lot so the tests fail. latest will always be there

File: kin/blockchain/channel_manager.py [hunk 4]
```diff
@@ -2,95 +2,84 @@
...
     TAKEN = 'taken'
     UNDERFUNDED = 'underfunded'
 
 
-# TODO: remove object when we kill python2
-class ChannelPool(queue.Queue, object):
+class ChannelPool(queue):
     """
-    A thread-safe queue that sets a member's status instead of pulling it in/out of the queue.
+    An async queue that sets a member's status instead of pulling it in/out of the queue.
```
Comment (yosriz) [inline kin/blockchain/channel_manager.py]: out of curiosity, why queue is needed? seems simple list/set will provide the same functionality? I mean there's no real use of push/pop here.
(mainly trying to understand that I'm not missing something with my node impl for that)

File: kin/monitors.py [hunk 5]
```diff
@@ -1,227 +1,68 @@
...
-        self.thread.daemon = True
-        self.thread.start()
-
-    def event_processor(self, stop_event):
-        """
-        Method to filter through SSE events and find kin payments for an account
-        :param threading.Event stop_event: an event that can be used to stop this method
-        """
-        import json
+async def single_monitor(kin_client: 'KinClient', address: str,
```
Comment (yosriz) [inline kin/monitors.py]: why does the KinClinet is wrapped with ' ?

File: kin/monitors.py [hunk 6]
```diff
@@ -1,227 +1,68 @@
...
+    Monitors a single account for kin payments
+
+    :param kin_client: a kin client directed to the correct network
+    :param address: address to watch
+    :param timeout: How long to wait for a new event
+
+    :raises: asyncio.TimeoutError: If too much time has passed between events (only if "timeout" is set)
+    """
+    if not is_valid_address(address):
+        raise StellarAddressInvalidError('invalid address: {}'.format(address))
```
Comment (yosriz) [inline kin/monitors.py]: can we remove `Stellar` from this error name?

File: kin/monitors.py [hunk 7]
```diff
@@ -1,227 +1,68 @@
...
-
-    def event_processor(self, stop_event):
-        """
-        Method to filter through SSE events and find kin payments for an account
-        :param threading.Event stop_event: an event that can be used to stop this method
-        """
-        import json
+            tx_data = SimplifiedTransaction(RawTransaction(tx))
+        except CantSimplifyError:
+            logger.debug("SSE transaction couldn't be simplified: ", tx)
```
Comment (yosriz) [inline kin/monitors.py]: what's exactly this logger? is it controllable by the developer?

File: kin/monitors.py [hunk 8]
```diff
@@ -1,227 +1,68 @@
...
+            continue
+
+        if tx_data.operation.type != OperationTypes.PAYMENT:
+            logger.debug("Non-payment SSE transaction skipped: ", tx_data)
+            continue
+
+        # Will yield twice if both of these are correct. (someone sent to himself) - which it fine
```
Comment (yosriz) [inline kin/monitors.py]: why not check both and send an only single event? (not crucial as it's an edge case, but don't see a reason to fire twice)

File: test/test_monitor.py [hunk 9]
```diff
@@ -1,85 +1,10 @@
...
-    sleep(5)
-    assert hash3 in txs_found2
-
-    # stop monitoring
-    monitor.stop()
-    hash4 = test_account.send_kin(address2, 4, fee=100)
-    sleep(10)
-    assert not monitor.thread.is_alive()
-    assert hash4 not in txs_found2
+    pass
```
Comment (yosriz) [inline test/test_monitor.py]: what exactly is broken? does it happen also on testnet?

File: kin/client.py [hunk 10]
```diff
@@ -172,24 +183,26 @@ def get_transaction_data(self, tx_hash, simple=True):
...
+            raw_tx = RawTransaction(await self.horizon.transaction(tx_hash))
         except Exception as e:
             raise KinErrors.translate_error(e)
 
         if simple:
             return SimplifiedTransaction(raw_tx)
         return raw_tx
 
-    def get_account_tx_history(self, address, amount=10, descending=True, cursor=None, simple=True):
+    async def get_account_tx_history(self, address: str, amount: Optional[int] = 10, descending: Optional[bool] = True,
```
[the remark this answers is not in the document]
Comment (Ronserruya) [reply kin/client.py]: @mrsegev This has been exposed for months now, should I remove it then?

File: kin/client.py [hunk 11]
```diff
@@ -279,49 +258,32 @@ def friendbot(self, address):
...
         if not is_valid_address(address):
             raise KinErrors.StellarAddressInvalidError('invalid address: {}'.format(address))
-        if self.does_account_exists(address):
+        if await self.does_account_exists(address):
             raise KinErrors.AccountExistsError(address)
 
-        response = requests.get(self.environment.friendbot_url, params={'addr': address})
-        if response.ok:
-            return response.json()['hash']
+        response = await self.horizon._session.get(self.environment.friendbot_url, params={'addr': address})
```
[the remark this answers is not in the document]
Comment (Ronserruya) [reply kin/client.py]: Good idea, I will create a ticket

File: kin/blockchain/channel_manager.py (hunk 4 above)
Comment (Ronserruya) [reply kin/blockchain/channel_manager.py]: There is really no use of any push/pop here. I used the queue class before since it helped me with thread-safety. 

Now I'm inheriting it since it has asnyc methods to get/push back to the "queue". Otherwise, I would need to write my own implementation to select an item from the set asyncrounsly

File: kin/monitors.py (hunk 5 above)
Comment (Ronserruya) [reply kin/monitors.py]: Kin client imports "Single Monitor"

If I wanted to write KinClient without '' I would need to import KinClient in SingleMonitor, and then I would have a circular loop of importing. So instead of importing the type you can just write it as a string. That's called 'forward referencing'

https://stackoverflow.com/questions/39740632/python-type-hinting-without-cyclic-imports

File: kin/monitors.py (hunk 7 above)
Comment (Ronserruya) [reply kin/monitors.py]: Just a logger like any other. And yes, you can set it's level

See https://github.com/kinecosystem/migration-server/blob/master/scripts/src/init.py#L42

File: kin/monitors.py (hunk 8 above)
Comment (Ronserruya) [reply kin/monitors.py]: This is just how it worked so far so I didnt wanted to change it

File: test/test_monitor.py (hunk 9 above)
Comment (Ronserruya) [reply test/test_monitor.py]: I tested my sse implementation manually and it works, but yes we get these 404s from time to time so I didn't want to get my test to fail because of a BC issue

File: kin/blockchain/channel_manager.py (hunk 4 above)
Comment (yosriz) [reply kin/blockchain/channel_manager.py]: ok, so I understood it correctly, I would suggest to call it `pool` instead of `queue` (just in case it is a public api)

File: kin/monitors.py (hunk 5 above)
Comment (yosriz) [reply kin/monitors.py]: python is fun!
